packages/ui/src/components/Channel/Scroller.tsx

```diff
@@ -169,62 +169,27 @@ const Scroller = forwardRef(
       [handleSetActive]
     );
 
-    const userHasScrolledRef = useRef(false);
-    const renderedPostsRef = useRef(new Set());
-    // Whether we've scrolled to the anchor post.
-    const [hasFoundAnchor, setHasFoundAnchor] = useState(!anchor);
-
-    logger.log('anchor postId', anchor?.postId);
-    const anchorIndex = useMemo(() => {
-      return posts?.findIndex((p) => p.id === anchor?.postId) ?? -1;
-    }, [posts, anchor]);
-
-    logger.log('anchorIndex', anchorIndex, anchor?.postId);
-    logger.log('hasFoundAnchor', hasFoundAnchor);
-
-    // We use this function to manage autoscrolling to the anchor post. We need
-    // the post to be rendered before we're able to scroll to it, so we wait for
-    // it here. Once it's rendered we scroll to it and set `hasFoundAnchor` to
-    // true, revealing the Scroller.
-    const handleItemLayout = useCallback(
-      (post: db.Post, index: number) => {
-        renderedPostsRef.current.add(post.id);
-        if (
-          !userHasScrolledRef.current &&
-          (post.id === anchor?.postId || (hasFoundAnchor && anchorIndex !== -1))
-        ) {
-          logger.log('scrolling to initially set anchor', post.id, index);
-          flatListRef.current?.scrollToIndex({
-            index: anchorIndex,
-            animated: false,
-            viewPosition: anchor?.type === 'unread' ? 1 : 0.5,
-          });
-        }
-
-        if (
-          !hasFoundAnchor &&
-          (anchor?.postId === post.id ||
-            // if we've got at least a page of posts and we've rendered them all,
-            // reveal the scroller to prevent getting stuck when messages are
-            // deleted.
-            (posts?.length && renderedPostsRef.current.size >= posts?.length))
-        ) {
-          logger.log('found anchor', post.id, index);
-          setHasFoundAnchor(true);
-        }
-      },
-      [anchor?.postId, anchorIndex, hasFoundAnchor, posts?.length, anchor?.type]
-    );
+    const {
+      readyToDisplayPosts,
+      scrollerItemProps: anchorScrollLockScrollerItemProps,
+      flatlistProps: anchorScrollLockFlatlistProps,
+    } = useAnchorScrollLock({
+      posts,
+      anchor,
+      flatListRef,
+      hasNewerPosts,
+      channelType,
+    });
 
     const theme = useTheme();
 
     // Used to hide the scroller until we've found the anchor post.
     const style = useMemo(() => {
       return {
-        opacity: hasFoundAnchor ? 1 : 0,
+        opacity: readyToDisplayPosts ? 1 : 0,
```

Comment:
I think this is firing too frequently. When searching for results far in scrollback and scrolling down, I'm noticing a blank screen/scroll thrash immediately after loading newer posts.

Reply:
chatted in DM – I'm seeing this as well when jumping to an older message via search -> tap, then scrolling down towards newer messages – I agree that this is a blocker, and it's proving to be tricky, so I thought I'd do a lil check in + notes:

- This doesn't appear to be related to `autoscrollToTopThreshold`, as `maintainVisibleContentPositionConfig` does not change after user scrolls for the first time.
- Posts are loading in as expected (+50 posts to the appropriate side with no stray null `posts`)
- We hide the scroller until it's ready to show, so I thought there might be a bug there – that appears to be working as expected

My current hypothesis is that this is due to `FlatList`'s async rendering combined with `maintainVisibleContentPositionConfig`. During the bug, the scroll view is technically scrolling >5000pts instantaneously:
1. We're adding 50 new posts once we hit the bottom of the scroll
2. which grows the scroll content size
3. which naturally changes what content the viewport is looking at
4. `maintainVisibleContentPositionConfig` jumps that 5000px to move the viewport back to the content we were originally looking at

Since `FlatList` discards offscreen content to make way for onscreen content, I think we rapidly discarding old content for new content, then immediately jumping back to that "older" content and redrawing it.

I don't know why this wasn't happening before (confirmed that the current released client does not have this issue).

Reply:
total gamer moment, looks like this is happening on `develop`, and it is also happening on release client (although it's faster on release).